dvc/remote/base.py

```diff
@@ -630,6 +630,9 @@ def exists_with_progress(chunks):
         return list(set(checksums) & set(self.all()))
 
     def already_cached(self, path_info):
+        if not self.cache:
```

Comment:
why this change? was there a bug before?

Reply:
`output.remote` doesn't have a `path_info` and crashes the whole thing, the alternative was to use `output.cache.already_cached` (as `cache` has `path_info`), however, when a remote is not configured (for example, non-cached external outputs) `cache` would be None.

Remembering these edge cases is error prone, so now you can use `output.remote.already_cached` and you'll be covered.

Reply:
so, what is the difference with was here before? I'm not sure I understand this still.

Reply:
We are calling `stage.remove_outs` on `Stage.create` (we are deleting the outputs to make sure the command is generating them)

```python
        # dvc/stage.py (create)

        # Removing outputs for reproducibility sake,
        # it is expected that the `cmd` will create them.
        if stage.cmd:
            confirm = kwargs.get("remove_outs", False)
            stage.remove_outs(ask=not confirm)
```

I added the `ask` parameter to `remove_outs` (previously, it just removed the outputs without asking (calling `output.remove`), but now, `output.remove` calls `output.safe_remove` if `ask` is set to `true` (this method asks the user for confirmation before it deletes the output).

```python
    # dvc/stage.py
    def remove_outs(self, ignore_remove=False, force=False, ask=False):
        """Used mainly for `dvc remove --outs` and :func:`Stage.reproduce`."""
        for out in self.outs:
            if out.persist and not force:
                out.unprotect()
            else:
                logger.debug(
                    "Removing output '{out}' of '{stage}'.".format(
                        out=out, stage=self.relpath
                    )
                )

                if force:
                    ask = False

                out.remove(ignore_remove=ignore_remove, force=not ask)
```

```python
    # dvc/output/base.py
    def remove(self, ignore_remove=False, force=True):
        self.remote.safe_remove(self.path_info, force=force)
```

This is were things start to get confusing with the `ask`, `force`, and `confirm`.

* For `stage.remove_outs`:
  * `force` means to remove persisting outputs.
  * `ask` means to _ask for confirmation_ when removing outputs (by default set to false (to not affect the previously defined interface)

For `output.remove`:
  * `force` is passed directly to `safe_remove`, and it means to not ask for confirmation while "safe removing".

Now, `output.remove` calls `remote.safe_remove`.
For `safe_remove` to work, it needs a `cache_dir` (because safe remove checks if the output is on the cache, if it has no cache it crashes -- this was not a bug before because we were not using safe remove on `stage.create` but now, if the remote has no `cache_dir` it will break, that's why I'm checking if the cache exists and sending methods to the `cache` object instead of the `remote` one).

Imagine the following case:

```bash
dvc remote add ssh ssh://localhost
dvc run --outs-no-cache remote://ssh/file.txt
```

The stage file is going to have one output.
This output doesn't require any cache.

This is where the `if not self.cache` is relevant.

I hope this helps, it took me a while to get everything on my head.